specs/experimental/gov-proposal-validator.md

```diff
@@ -0,0 +1,581 @@
+function submitFundingProposal(
+    uint128 _criteriaValue,
+    string[] _optionsDescriptions,
+    address[] _optionsRecipients,
+    uint256[] _optionsAmounts,
+    string memory _description,
+    ProposalType _proposalType,
+) external returns (bytes32 proposalHash_);
+```
+
+Approval Voting Module
+
+Funding proposals use the `Approval` voting module but unlike the Protocol or Governor upgrade proposals,
+funding proposals need to execute token transfers.
+This requires the user who submits the proposal to provide some additional data related to the proposal.
+
+For the `ProposalSettings` of the voting module, these are:
+- `uint128 criteriaValue`: Since the passing criteria type is always "Threshold" for this proposal type
+this will be the amount of votes that the proposal need to gather to pass.
+
+For the `ProposalOptions` of the voting module, these are:
+- `string[] optionsDescriptions`: The strings of the different options that can be voted.
+- `address[] optionsRecipients`: An address for each option to transfer funds to in case the option passes the voting.
+- `uint256[] optionsAmounts`: The amount to transfer for each option in case the option passes the voting.
+
+`approveProposal`
+
+Approves a proposal before being moved for voting, used by the top delegates.
+
+- MUST check if proposal hash corresponds to a valid proposal
+- MUST check if caller has enough voting power to call the function and approve a proposal
+  The voting power of the delegate is checked against the end of the last voting cycle from when
+  the proposal was submitted.
+- MUST check if caller has already approved the same proposal
+- MUST store the approval vote
+- MUST emit `ProposalApproved` when successfully called
+
+```solidity
+function approveProposal(bytes32 _proposalHash) external
+```
+
+Approving a funding proposal type requires extra attention to the budget amount and options, of the
+approval voting module, that were provided on the submission of the proposal. This should be handled
+by the Agora's UI.
+
+`moveToVote`
+
+Checks if the provided proposal is ready to move for voting. Based on the Proposal Type different checks are being
+validated. If all checks pass then `OptimismGovernor.proposeWithModule` is being called to forward the proposal for voting.
+
+- MUST create the `proposalHash` and check if it exists and is valid
+- Proposal MUST have gathered equal or more than the `requiredApprovals` defined for that type
+- MUST check if proposal has already moved for voting
+- MUST emit `ProposalMovedToVote` event
+
+For `ProtocolOrGovernorUpgrade` type:
+
+- MUST also check that is called by the proposer that submitted the proposal
+- `_optionsRecipients` and `_optionsAmounts` MUST be empty
+
+For `MaintenanceUpgradeProposals` type:
+
+- This type does not require any checks and is being forwarded to the Governor contracts, this should happen atomically.
+
+For `CouncilMemberElections`, `GovernanceFund` and `CouncilBudget` types:
```

Comment:
Elections do not do fund transfers.  So I believe that the `_optionsAmounts/recipients` should be checked as zero in the `CouncilMemberElections` case, same as upgrades case

Reply:
Correct, for the `CouncilMemberElections` we will be checking to make sure these fields are empty, will  update the document to mention this